kratos/geometries/nurbs_volume_geometry.h

```diff
@@ -700,69 +734,160 @@ class NurbsVolumeGeometry : public Geometry<typename TContainerPointType::value_
-            mPolynomialDegreeU, mPolynomialDegreeV, mPolynomialDegreeW, NumberOfShapeFunctionDerivatives);
+        // Option 1: One QuadraturePointGeometry is created containing all integration points. This should be used
+        // when all rIntegrationPoints are located inside the same element.
+        if(  IntegrationInfo::QuadratureMethod::GAUSS == rIntegrationInfo.GetQuadratureMethod(0) ){
+            KRATOS_ERROR_IF(NumberOfShapeFunctionDerivatives != 2) << "NumberOfShapeFunctionDerivatives must be 2.\n";
+            // Shape function container.
+            NurbsVolumeShapeFunction shape_function_container(
+                mPolynomialDegreeU, mPolynomialDegreeV, mPolynomialDegreeW, NumberOfShapeFunctionDerivatives);
+
+            // Resize containers.
+            if (rResultGeometries.size() != 1){
+                rResultGeometries.resize(1);
+            }
 
-        // Resize containers.
-        if (rResultGeometries.size() != rIntegrationPoints.size())
-            rResultGeometries.resize(rIntegrationPoints.size());
+            const auto default_method = this->GetDefaultIntegrationMethod();
 
-        auto default_method = this->GetDefaultIntegrationMethod();
-        SizeType num_nonzero_cps = shape_function_container.NumberOfNonzeroControlPoints();
+            const SizeType num_nonzero_cps = shape_function_container.NumberOfNonzeroControlPoints();
+            const SizeType num_points = rIntegrationPoints.size();
+            KRATOS_ERROR_IF(num_points < 1) << "List of integration points is empty.\n";
 
-        Matrix N(1, num_nonzero_cps);
-        DenseVector<Matrix> shape_function_derivatives(NumberOfShapeFunctionDerivatives - 1);
+            // Initialize containers.
+            IntegrationPointsContainerType integration_points;
+            ShapeFunctionsValuesContainerType shape_function_values;
+            ShapeFunctionsLocalGradientsContainerType shape_function_gradients;
 
-        for (IndexType i = 0; i < NumberOfShapeFunctionDerivatives - 1; ++i) {
-            const IndexType num_derivatives = (2 + i) * (3 + i) / 2;
-            shape_function_derivatives[i].resize(num_nonzero_cps, num_derivatives);
-        }
+            integration_points[0] = rIntegrationPoints;
+            shape_function_gradients[0].resize(rIntegrationPoints.size());
+            shape_function_values[0].resize(rIntegrationPoints.size(), num_nonzero_cps);
 
-        for (IndexType i = 0; i < rIntegrationPoints.size(); ++i)
-        {
-            // Attention: Weights are not yet implemented.
-            // if (IsRational()) {
-            //     shape_function_container.ComputeNurbsShapeFunctionValues(
-            //         mKnotsU, mKnotsV, mKnotsW, mWeights, rIntegrationPoints[i][0], rIntegrationPoints[i][1], rIntegrationPoints[i][2]);
-            // }
+            for( IndexType i_point = 0; i_point < num_points; ++i_point){
+                for (IndexType i = 0; i < NumberOfShapeFunctionDerivatives - 1; ++i) {
```

Comment:
`NumberOfShapeFunctionDerivatives` is guaranteed to be 2, right? Then everything in the loop scope runs for `i=0` only and doesn´t need the loop at all.

Reply:
It is not always. In the input, we assign the shape function order, which can be 1 or 2 and so on (case dependent).

Reply:
but there's a check in L740 at the head of this branch:
```cpp
KRATOS_ERROR_IF(NumberOfShapeFunctionDerivatives != 2) << "NumberOfShapeFunctionDerivatives must be 2.\n";
```

Reply:
My bad. Yes, in this branch, we can omit the loop.

Reply:
Yes, that is correct. Actually, the current interface does not provide the necessary containers for the higher-order derivatives. I thought I keep it in case we change it at some point. I could comment it out. In fact, for the solids we most likely never require second-order derivatives..

Reply:
The problem is that the current interface does not support what I want to do. Therefore, this is a bit hacky. I think in the long term we should think about changing the entire interface. However, if this is not important for the shells etc. I guess it is not worth the effort.

Reply:
It's OK to keep the loop if you're planning to add support for other orders later on; I just though there's an algorithmic restriction to `2`.